bson/bson_test.go

```diff
@@ -112,6 +117,37 @@ func TestD(t *testing.T) {
 	})
 }
 
+type stringerString string
+
+func (ss stringerString) String() string {
+	return "bar"
+}
+
+func TestMapCodec(t *testing.T) {
+	t.Run("EncodeKeysWithStringer", func(t *testing.T) {
```

Comment:
So, what about decoding... My assertion here is that this doesn't work because when we read it back we get "bar", which is decidedly _not_ "foo".  

In the ticket, this isn't a problem because they are encoding them as the underlying string value, not the result of the String() function. But if the type is an integer, we'll be writing it as a string and then decoding it as a string to assign it to an int? Do we handle that already?

Anyways, perhaps you need a 
```
type KeyCodec interface {
  Encode(val reflect.Value) (string, error)
  Decode(val string) (reflect.Value, error)
}
```

Reply:
As discussed, the default behavior was switched to the encoding/json behavior, which can be roundtripped